pageserver/src/walreceiver/walreceiver_connection.rs

```diff
@@ -315,18 +315,20 @@ pub async fn handle_walreceiver_connection(
 
             // Send zenith feedback message.
             // Regular standby_status_update fields are put into this message.
-            let zenith_status_update = ReplicationFeedback {
-                current_timeline_size: timeline.get_current_logical_size() as u64,
+            let status_update = ReplicationFeedback {
+                current_timeline_size: timeline
+                    .get_current_logical_size()
```

Comment:
This is now the only place where we actually init the timeline logical size.
This could have changed the performance due to the way we calculated the logical size before and due to the fact that we used to load timeline everywhere, so the size init could have happened before the walreceiver process had to send the feedback.

Now, we do that here and it seems that tests are fine, yet I think we could improve it (maybe, calculate the size in the background on walreceiver start?)

Reply:
Yeah, would be nice to calculate it in the background. As it is, this blocks digesting new WAL.

Hmm, come to think of it, if we call get_current_logical_size() in the walreceiver, doesn't that render the whole lazy calculation pointless? The only thing that ingests new WAL to the timeline, and updates the logical size, is the WAL receiver. If the WAL receiver is also the only thing that performs the calculation, we can never be performing the initial calculation and digesting new WAL at the same time.

P.S. It actually feels weird to me that we send the logical size to the safekeepers, and the safekeeper forwards it to the compute node. It would feel more natural to send it directly from the pageserver to the compute node. But we don't have a feedback channel to send updates from the pageserver to the compute node, currently. Would need some redesigning of the protocol. (I don't suggest doing that in this PR, but maybe later)

Reply:
I've adjusted the code to run a single background thread per timeline to calculate its logical size.

I'm not super happy with the current impl due to obvious reasons:
* it's not a timeline's responsibility to run and orchestrate tasks, yet it's doing that
* a task is run in a separate thread without any cancellation and the managing code around is somewhat dirty compared to async or rust 1.61

I think those all should be fixed when we agree on the sync-async more and define better our task manager, the latter should be on my list next week.

Reply:
I was wrong, the code is adjusted and works now.